src/osd/OSD.cc

```diff
@@ -6264,23 +6265,42 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
 
   int ret = smartctl.spawn();
   if (ret != 0) {
-    derr << "failed run smartctl: " << smartctl.err() << dendl;
-    return ret;
+    std::string err_msg = "failed to run smartctl: " + smartctl.err();
+    derr << err_msg << dendl;
+    *result = err_msg;
+    return -1;
   }
 
   bufferlist output;
   ret = output.read_fd(smartctl.get_stdout(), 100*1024);
   if (ret < 0) {
-    derr << "failed read from smartctl: " << cpp_strerror(-ret) << dendl;
+    std::string err_msg = "failed to read from smartctl: " + cpp_strerror(-ret);  
+    derr << err_msg << dendl;
+    *result = err_msg;
   } else {
-    ret = 0;
     *result = output.to_str();
     dout(10) << "smartctl output is: " << *result << dendl;
   }
 
-  if (smartctl.join() != 0) {
+  ret = smartctl.join();
+  /*
+  The exit statuses of smartctl are defined by a bitmask.
+  if (ret == 1) : 
+    a. smartctl returned 1, which means command line did not parse.
+    b. there might be a sudoers permission problem;
+  if (ret == 2) :
+    smartctl returned 2 which means: 
+    device open failed,
+    device did not return an IDENTIFY DEVICE structure,
+    or device is in a low-power mode.
+  (ret == 3) is not expected (due to bitmask).
+  The rest of smartctl exit status suggest there might be something wrong with the device,
+  but the output might still be useful. When (0 < ret < 4)
+  there is no actual output about the device (such as attr table, etc.), 
+  and we need to check what went wrong.
+  */
+  if (ret != 0) {
     derr << smartctl.err() << dendl;
-    return -EINVAL;
```

Comment:
I think this is the only piece that needs to change. Maybe add in here that if result is empty (smartctl returned error *and* failed to produce some output) then we set result to { "error": "smartctl returned error", "exit_code": r } or something.